### PR TITLE
ibmcloud: e2e test harden cluster subnet and vpc

### PR DESCRIPTION
**Pull request by huoqifeng**
Fixes: #689

### CODE REVIEW

**Comment by huoqifeng** [comment]
In this PR, added logic to check existing for cluster, vpc and subnet before creating and deleting. Corresponding IDs are not required and deleted. Another major change is to use package github.com/IBM/vpc-go-sdk `v0.32.0` rather than `v1.0.1` as consulted Ujjwal Kumar(https://github.com/ujjwal-ibm),  `v0.32.0` is actively maintained version, which is also the latest release. So, @yoheiueda would you please check whether the change in `provider.go` good for you.

**Comment by huoqifeng** [comment]
To run the test:
```
# make build

# export LOG_LEVEL="debug"
# export CLOUD_PROVIDER="ibmcloud"
# export TEST_E2E_PROVISION_FILE="/root/provision_ibmcloud.properties"
# export TEST_E2E_PODVM_IMAGE="/root/e2e-test-image-amd64-20230308-1.qcow2"
# export TEST_E2E_PROVISION="yes"

# ./cluster-provisioner -action=provision
DEBU[0000] &{ApiKey:xxx Bucket:peerpod-cos-bucket ClusterName:e2e-test2 CosApiKey:xxx CosInstanceID:crn:v1:bluemix:public:cloud-object-storage:global:a/cdd0a83a81fd38ed101096a049942e6d:1bd99fa3-2a6b-45f7-815d-2271033fdaa5:: CosServiceURL:s3.jp-tok.cloud-object-storage.appdomain.cloud SecurityGroupID: IamServiceURL:https://iam.cloud.ibm.com/identity/token InstanceProfile:bz2-2x8 PodvmImageID: PodvmImageArch:s390x Region:jp-tok ResourceGroupID:33e11ea6cbf340f2bfb6c8414f511b8d SshKeyID:r022-3ec05389-9662-44a9-8f84-71edde2e6af3 SubnetName:e2e-test2-vpc-subnet SubnetID: VpcName:e2e-test2-vpc VpcID: VpcServiceURL:https://jp-tok.iaas.cloud.ibm.com/v1 WorkerFlavor:bx2.2x8 Zone:jp-tok-1 WorkerCount:1 IsSelfManaged:false VPC:<nil> ClusterAPI:<nil>} 
INFO[0002] Creating VPC e2e-test2-vpc in ResourceGroupID 33e11ea6cbf340f2bfb6c8414f511b8d. 
INFO[0006] Created VPC with ID r022-055991f8-9406-496e-8d01-ffe0b55917e6 in ResourceGroupID 33e11ea6cbf340f2bfb6c8414f511b8d. 
INFO[0007] Got VPC default SecurityGroupID r022-fe0c2a48-acb4-4a2e-b856-e69d9f3945dc. 
INFO[0007] Waiting for the VPC to be available before creating subnet... 
INFO[0067] Creating subnet e2e-test2-vpc-subnet in VPC r022-055991f8-9406-496e-8d01-ffe0b55917e6 in Zone jp-tok-1. 
INFO[0070] Created subnet with ID 02e7-41feb296-63bf-44d4-913d-f28180ae6a13. 
INFO[0071] Creating cluster e2e-test2.                  
INFO[0080] Waiting for cluster e2e-test2 to be available. 
INFO[0080] Waited 0 minutes...                          
INFO[0141] Waited 1 minutes...                          
INFO[0202] Waited 2 minutes...                          
INFO[0262] Waited 3 minutes...                          
INFO[0323] Waited 4 minutes...                          
INFO[0384] Waited 5 minutes...                          
INFO[0444] Waited 6 minutes...                          
INFO[0505] Waited 7 minutes...                          
INFO[0565] Waited 8 minutes...                          
INFO[0626] Waited 9 minutes...                          
INFO[0687] Waited 10 minutes...                         
INFO[0747] Waited 11 minutes...                         
INFO[0808] Waited 12 minutes...                         
INFO[0868] Waited 13 minutes...                         
INFO[0929] Waited 14 minutes...                         
INFO[0990] Waited 15 minutes...                         
INFO[1050] Waited 16 minutes...                         
INFO[1111] Cluster e2e-test2 is available.     


# ./cluster-provisioner -action=deprovision
DEBU[0000] &{ApiKey:xxx Bucket:peerpod-cos-bucket ClusterName:e2e-test2 CosApiKey:xxx CosInstanceID:crn:v1:bluemix:public:cloud-object-storage:global:a/cdd0a83a81fd38ed101096a049942e6d:1bd99fa3-2a6b-45f7-815d-2271033fdaa5:: CosServiceURL:s3.jp-tok.cloud-object-storage.appdomain.cloud SecurityGroupID: IamServiceURL:https://iam.cloud.ibm.com/identity/token InstanceProfile:bz2-2x8 PodvmImageID: PodvmImageArch:s390x Region:jp-tok ResourceGroupID:33e11ea6cbf340f2bfb6c8414f511b8d SshKeyID:r022-3ec05389-9662-44a9-8f84-71edde2e6af3 SubnetName:e2e-test2-vpc-subnet SubnetID: VpcName:e2e-test2-vpc VpcID: VpcServiceURL:https://jp-tok.iaas.cloud.ibm.com/v1 WorkerFlavor:bx2.2x8 Zone:jp-tok-1 WorkerCount:1 IsSelfManaged:false VPC:<nil> ClusterAPI:<nil>} 
INFO[0001] Deleting Cluster e2e-test2.                  
INFO[0003] Waiting for cluster e2e-test2 to be removed... 
INFO[0003] Waited 0 minutes...                          
INFO[0064] Waited 1 minutes...                          
INFO[0124] Waited 2 minutes...                          
INFO[0185] Waited 3 minutes...                          
INFO[0245] Cluster e2e-test2 is removed.                
INFO[0248] Found subnet e2e-test2-vpc-subnet with ID 02e7-41feb296-63bf-44d4-913d-f28180ae6a13. 
INFO[0248] Deleting subnet with ID 02e7-41feb296-63bf-44d4-913d-f28180ae6a13. 
INFO[0250] Deleted subnet with ID 02e7-41feb296-63bf-44d4-913d-f28180ae6a13. 
INFO[0251] Found VPC e2e-test2-vpc with ID r022-055991f8-9406-496e-8d01-ffe0b55917e6. 
INFO[0251] Deleting VPC with ID r022-055991f8-9406-496e-8d01-ffe0b55917e6. 
INFO[0251] Deleted VPC with ID r022-055991f8-9406-496e-8d01-ffe0b55917e6. 

```

File: test/provisioner/provision_ibmcloud.properties [hunk 1]
```diff
@@ -12,26 +12,20 @@ COS_APIKEY="${MY_COS_SERVICE_APIKEY}"
...
 WORKER_FLAVOR="bx2.2x8"
 WORKERS_COUNT="1"
-ZONE="jp-tok-1"
\ No newline at end of file
+ZONE="jp-tok-1"
+# optional, required only when use existing podvm image but not upload new image
+PODVM_IMAGE_ID=""
+# optional, it'll be set as ${CLUSTER_NAME}-vpc-subnet if not provided
+VPC_SUBNET_NAME=""
+# optional, it can be set as ${CLUSTER_NAME}-vpc if not provided
```
Comment (liudalibj) [inline test/provisioner/provision_ibmcloud.properties:30]: `it can be` => `it'll be`

File: test/provisioner/provision_ibmcloud.go [hunk 2]
```diff
@@ -272,20 +295,73 @@ func isClusterReady(clrName string) (bool, error) {
...
+	for _, vpc := range allResults {
+		log.Tracef("Checking vpc %s.", *vpc.Name)
+		if *vpc.Name == vpcName {
+			return &vpc, nil
+		}
+	}
+	return nil, nil
+}
+
+// TODO, nice to have rety if SDK client did not do that for well known http errors
```
Comment (liudalibj) [inline test/provisioner/provision_ibmcloud.go:341]: `rety` => `retry`?

File: test/provisioner/provision_ibmcloud.go [hunk 3]
```diff
@@ -272,20 +295,73 @@ func isClusterReady(clrName string) (bool, error) {
 	return false, nil
 }
 
-func foundCluster(clrName string) (bool, error) {
+// TODO, nice to have rety if SDK client did not do that for well known http errors
```
Comment (liudalibj) [inline test/provisioner/provision_ibmcloud.go:298]: `rety` => `retry`?

**Review by liudalibj — APPROVED** [review]
there are some typo in the comments, others LGTM.

File: test/provisioner/provision_ibmcloud.go (hunk 3 above)
Comment (huoqifeng) [reply test/provisioner/provision_ibmcloud.go]: Fixed

**Review by stevenhorsman — APPROVED** [review]
Thanks Qifeng - this looks good and I was able to provision a vpc, subnet, cluster just based on the cluster name and then deprovision removed all of them!

**Pull request merged**